cmd/fts/cmd/run.go

```diff
@@ -0,0 +1,170 @@
+	runCmd.Flags().Int64Var(&minDealBytes, "min-deal-bytes", 0, "Minimum bytes of a single deal")
+	runCmd.Flags().MarkHidden("concurrent")
+	runCmd.Flags().MarkHidden("max-staged-bytes")
+	runCmd.Flags().MarkHidden("max-deal-bytes")
+	runCmd.Flags().MarkHidden("min-deal-bytes")
+
+	rootCmd.AddCommand(runCmd)
+}
+
+var runCmd = &cobra.Command{
+	Use:   "run",
+	Short: "Run a storage task pipeline",
+	Long:  `Run a storage task pipeline to migrate large collections of data to Filecoin`,
+	PreRun: func(cmd *cobra.Command, args []string) {
+		err := viper.BindPFlags(cmd.Flags())
+		checkErr(err)
+		// set default variables
+		if mainnet {
+			if maxStagedBytes == 0 {
+				maxStagedBytes = 26843545600 // 25Gib
+			}
+			if maxDealBytes == 0 {
+				maxDealBytes = 4294967296 // 4Gib
+			}
+			if minDealBytes == 0 {
+				minDealBytes = 67108864 // 0.5Gib
+			}
+		} else { // localnet
+			if maxStagedBytes == 0 {
+				maxStagedBytes = 524288
+			}
+			if maxDealBytes == 0 {
+				maxDealBytes = 524288
+			}
+			if minDealBytes == 0 {
+				minDealBytes = 9000
+			}
+		}
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		if maxStagedBytes < maxDealBytes {
+			Fatal(fmt.Errorf("Max deal size (%d) is larger than max staging size (%d)", maxDealBytes, maxStagedBytes))
+		}
+
+		// Read all tasks from the input folder
+		unfiltered, err := pathToTasks(taskFolder)
+		checkErr(err)
+		Message("Input tasks: %d", len(unfiltered))
+		allTasks := filterTasks(unfiltered)
+
+		// Ensure there aren't existing tasks in the same queue
+		knownTasks, err := openResults(resultsOut)
+		checkErr(err)
+		Message("Existing tasks: %d", len(knownTasks))
+		allTasks = mergeNewTasks(knownTasks, allTasks)
+
+		pendingTasks := cleanErrors(removeComplete(allTasks), retryErrors)
+
+		Message("Pending tasks: %d", len(pendingTasks))
+		time.Sleep(1 * time.Second)
+
+		if len(pendingTasks) == 0 {
+			return
+		}
+
+		// Init or update our intermediate results
+		err = storeResults(resultsOut, allTasks)
+		checkErr(err)
+
+		// Determine the max concurrent we need vs allowed
+		if int64(len(pendingTasks)) < concurrent {
+			concurrent = int64(len(pendingTasks))
+		}
+
+		rc := PipelineConfig{
+			token:           viper.GetString("token"),
+			serverAddress:   viper.GetString("serverAddress"),
+			ipfsrevproxy:    ipfsrevproxy,
+			maxStagingBytes: maxStagedBytes,
+			minDealBytes:    minDealBytes,
+			concurrent:      concurrent,
+			dryRun:          dryRun,
+		}
+		storageConfig, err := getStorageConfig(rc)
+		checkErr(err)
+		storageConfig.Hot.Enabled = false
+		rc.storageConfig = storageConfig
+
+		trackProgress(pendingTasks, allTasks, rc)
+	},
+}
+
+func trackProgress(pendingTasks []Task, allTasks []Task, conf PipelineConfig) {
+
+		var progressWaitGroup sync.WaitGroup
+
+		progress := uiprogress.New()
+		progress.Start()
+		progressBars := make(map[string](chan Task))
+
+		taskUpdates := Start(pendingTasks, conf)
+
+		j := 1
+		for task := range taskUpdates {
```

Comment:
all results are streamed back over this channel and all UI can be managed here. don't love the look right now, but didn't want to get hung up on it before people have tried this thing.